cli/src/pcluster/validators/database_validators.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+from urllib.parse import ParseResult, urlparse
+
+from pcluster.validators.common import FailureLevel, Validator
+
+
+class DatabaseUriValidator(Validator):
+    """Domain address validator."""
+
+    def _validate(self, uri: str):
+        """Validate database URI."""
+        # First, throw error if the URI starts with a "/" (to prevent issues with the
+        # manipulation below
+        if not self._check_trailing_slash(uri):
+            return
+
+        uri_parse = urlparse(uri)
+        if not uri_parse.netloc:
+            # This happens if users provide an URI without explicit scheme followed by ://
+            # (for example 'test.example.com:3306' instead of 'mysql://test.example.com:3306`).
+            uri_parse = urlparse("//" + uri)
+
+        # Throw error if the URI contains a scheme
+        if not self._check_scheme(uri_parse):
+            return
```

Comment:
For consistency in validator behaviour, should we avoid returning on the first encounter of an error?
I believe the current behaviour is to build a list of validation failures then print all of them.
Using this `return` approach will break that pattern.

Reply:
That's a very good question. When looking at the validators, most validation functions had a structure like `if ... elif ...`, meaning that they implicitly return only one type of validation error.

Some validators though show the pattern you are talking about (for example [`ExistingFsxNetworkingValidator`](https://github.com/aws/aws-parallelcluster/blob/874f95fac31731eae1b5156c5ca274c482d58cb4/cli/src/pcluster/validators/cluster_validators.py#L445)). But I believe there we are talking about looping through different entities to be validated (essentially one check per file system). Do you have any example about multiple validation errors for the same validated entity?

I'm not sure what the best user experience is. Should we tell the user all the criteria the string does not comply with, even if the string will most likely change to cope to the first validation error, potentially leading to a different outcome at the next validation iteration?

Reply:
> most validation functions had a structure like if ... elif ..., meaning that they implicitly return only one type of validation error.
This seems to happen when comparing a specific property as opposed to a range of properties.

For the intended experience, it may be worth discussing it with the wider team but it's not a technical blocker.

Reply:
We will leave it like this for the time being. Waiting for feedback from beta testers.